Zend/zend_max_execution_timer.c

```diff
@@ -41,6 +43,14 @@ ZEND_API void zend_max_execution_timer_init(void) /* {{{ */
 	sev.sigev_signo = SIGRTMIN;
 	sev.sigev_notify_thread_id = (pid_t) syscall(SYS_gettid);
```

Comment:
`zend_max_execution_timer_init` is called from `pcntl_fork`. With this change, a new timer will no longer be created. However, it seems `syscall(SYS_gettid)` will no longer match. This might be a problem.

Reply:
I'll create a test for this case to ensure we are not breaking it.

Reply:
I couldn't figure out how to write a phpt test for this case. I added some logging and a simple script to see what was going on though:

```php
<?php

echo "about to fork\n";
$pid = pcntl_fork();
if(!$pid) {
    echo sprintf("child process: %s \n", getmypid());
    sleep(1);
} else {
    echo sprintf("parent process: %s\n", getmypid());
    sleep(2);
}
```

And it produces the following output:

```
$ php -dmax_execution_time=10 test.php
Timer 0 created on thread 54661
Setting timer 0 on thread 54661 (10 seconds)...
about to fork
Timer 0 created on thread 54662
parent process: 54661
child process: 54662
Setting timer 0 on thread 54662 (0 seconds)...
Deleting timer 0 on thread 54662...
Setting timer 0 on thread 54661 (0 seconds)...
Deleting timer 0 on thread 54661...
```

So it appears this might just be defensive programming, since it seems a fork creates its timers (and maybe doesn't inherit the parent's settings?).

I can also see it handles max_execution_time correctly as well:

```php
<?php

echo "about to fork\n";
$pid = pcntl_fork();
if(!$pid) {
    ini_set('max_execution_time', 1);
    echo sprintf("child process: %s \n", getmypid());
    sleep(3);
} else {
    ini_set('max_execution_time', 4);
    echo sprintf("parent process: %s\n", getmypid());
    sleep(5);
}
```

which gives the following output:

```
$ php -dmax_execution_time=10 test.php
Timer 0 created on thread 54667
Setting timer 0 on thread 54667 (10 seconds)...
about to fork
Setting timer 0 on thread 54667 (0 seconds)...
Setting timer 0 on thread 54667 (4 seconds)...
parent process: 54667
Timer 0 created on thread 54668
Setting timer 0 on thread 54668 (0 seconds)...
Setting timer 0 on thread 54668 (1 seconds)...
child process: 54668
Setting timer 0 on thread 54668 (0 seconds)...
PHP Fatal error:  Maximum execution time of 1 second exceeded in /usr/local/src/php/test.php on line 5

Fatal error: Maximum execution time of 1 second exceeded in /usr/local/src/php/test.php on line 5
Setting timer 0 on thread 54668 (0 seconds)...
Deleting timer 0 on thread 54668...
Setting timer 0 on thread 54667 (0 seconds)...
PHP Fatal error:  Maximum execution time of 4 seconds exceeded in /usr/local/src/php/test.php on line 14

Fatal error: Maximum execution time of 4 seconds exceeded in /usr/local/src/php/test.php on line 14
Setting timer 0 on thread 54667 (0 seconds)...
Deleting timer 0 on thread 54667...
```

I'd love to write a test for this, but I'll have to dive into how phpt works to understand it. However, it looks like there may not be an issue here for pcntl_fork.

Reply:
I think switching to `EG(pid) == pid` implicitly fixed this. :+1: This could be tested by forking, making the child timeout, and monitoring the chlid from the parent. But I'm not sure a test is worth it.